src/librustc/mir/interpret/mod.rs

```diff
@@ -217,9 +218,56 @@ impl<'tcx, Tag> Pointer<Tag> {
 }
 
 
-#[derive(Copy, Clone, Eq, Hash, Ord, PartialEq, PartialOrd, Debug)]
+#[derive(Copy, Clone, Debug)]
 pub struct AllocId(pub u64);
 
+impl Eq for AllocId {}
+
+impl Hash for AllocId {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        ty::tls::with(|tcx| {
+            let alloc_type = tcx.alloc_map.lock().get(*self);
+            match alloc_type {
+                Some(alloc_type) => alloc_type.hash(state),
+                None => self.0.hash(state),
+            }
+        })
+    }
+}
+
+impl PartialEq for AllocId {
+    fn eq(&self, other: &Self) -> bool {
+        ty::tls::with(|tcx| {
+            let (s, o) = {
+                let map = tcx.alloc_map.lock();
+                (map.get(*self), map.get(*other))
```

Comment:
Uh, wait, what? That doesn't seem right to me.

Certainly, ptr comparison in miri should compare the actual IDs, the the content that's stored behind these IDs. What is the type of `s` and `o` here?

Reply:
This only applies to `AllocId`s that refer to global allocations. Comparing by id is inherently wrong. We never ran into any trouble, because we never used a `ConstValue` containing an `AllocId` in comparisons that mattered. Now that everything is `ByRef`, pattern match exhaustiveness checks failed because they couldn't figure out that two const patterns are the same if their `AllocId` differs.

Reply:
> This only applies to AllocIds that refer to global allocations. 

It is still grossly wrong in the context of the miri engine. Ptr equality wants to compare the IDs, not the allocations they point to, and the IDs should be newtyped to not mix them up with other integers.

> pattern match exhaustiveness checks failed because they couldn't figure out that two const patterns are the same if their AllocId differs.

Then it shouldn't use the same equality as the miri engine.  If you want to overwrite equality and hashing for `ConstValue`, I can support that, but doing it for a type used inside the engine makes me really nervous.